common/json_writer/src/lib.rs

```diff
@@ -22,46 +23,62 @@ pub enum JsonWriterError {
 
 impl JsonWriter {
     pub fn new() -> Self {
-        Self { buffer: Vec::new() }
+        Self {
+            buffer: Vec::new(),
+            needs_separator: false,
+        }
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
         Self {
             buffer: Vec::with_capacity(capacity),
+            needs_separator: false,
         }
     }
 
     pub fn write_key(&mut self, key: &str) -> Result<(), JsonWriterError> {
         self.write_str(key)?;
         self.buffer.push(b':');
+        self.needs_separator = false;
         Ok(())
     }
 
     pub fn write_str(&mut self, s: &str) -> Result<(), JsonWriterError> {
-        Ok(serde_json::to_writer(&mut self.buffer, s)?)
+        self.separate();
+        let () = serde_json::to_writer(&mut self.buffer, s)?;
+        self.needs_separator = true;
+        Ok(())
     }
 
     pub fn write_f64(&mut self, value: f64) -> Result<(), JsonWriterError> {
+        self.separate();
         match value.classify() {
             FpCategory::Normal | FpCategory::Zero | FpCategory::Subnormal => {
-                Ok(serde_json::to_writer(&mut self.buffer, &value)?)
+                let () = serde_json::to_writer(&mut self.buffer, &value)?;
+                self.needs_separator = true;
+                Ok(())
             }
             FpCategory::Infinite | FpCategory::Nan => {
                 Err(JsonWriterError::InvalidF64Value { value })
             }
         }
     }
 
-    pub fn write_separator(&mut self) {
-        self.buffer.push(b',');
+    fn separate(&mut self) {
+        if self.needs_separator {
+            self.buffer.push(b',');
+            self.needs_separator = false;
+        }
     }
 
     pub fn write_open_obj(&mut self) {
+        self.separate();
```

Comment:
I would not call `separate()` separate - between a key and a record value.

Reply:
And I would set `self.needs_separator = false`.

Reply:
`separate` really is `maybe_separate`. It always sets `needs_separator = false`.